caluma/caluma_core/types.py

```diff
@@ -66,44 +66,60 @@ class DjangoConnectionField(DjangoConnectionField):
     This can be removed, when (or better if)
     https://github.com/graphql-python/graphql-relay-py/issues/12
     is resolved.
-
-    TODO: properly implement max_limit, see
-    https://github.com/graphql-python/graphene-django/blob/b552dcac24364d3ef824f865ba419c74605942b2/graphene_django/fields.py#L133
     """
 
     @classmethod
     def resolve_connection(cls, connection, args, iterable, max_limit=None):
+        # Remove the offset parameter and convert it to an after cursor.
+        offset = args.pop("offset", None)
+        after = args.get("after")
+        if offset:  # pragma: no cover
+            if after:
+                offset += cursor_to_offset(after) + 1
+            # input offset starts at 1 while the graphene offset starts at 0
+            args["after"] = offset_to_cursor(offset - 1)
+
         iterable = maybe_queryset(iterable)
+
         if isinstance(iterable, QuerySet):
             # only query count on database when pagination is needed
             # resolve_connection may be removed again once following issue is fixed:
             # https://github.com/graphql-python/graphene-django/issues/177
             if all(args.get(x) is None for x in ["before", "after", "first", "last"]):
-                _len = len(iterable)
+                list_length = len(iterable)
             else:
-                _len = iterable.count()
+                list_length = iterable.count()
         else:  # pragma: no cover
-            _len = len(iterable)
+            list_length = len(iterable)
+        list_slice_length = (
+            min(max_limit, list_length) if max_limit is not None else list_length
+        )
 
         # If after is higher than list_length, connection_from_array_slice
         # would try to do a negative slicing which makes django throw an
         # AssertionError
-        after = min(get_offset_with_default(args.get("after"), -1) + 1, _len)
-        if max_limit is not None and "first" not in args:  # pragma: no cover
-            args["first"] = max_limit
+        after = min(get_offset_with_default(args.get("after"), -1) + 1, list_length)
+
+        if (
+            max_limit is not None and args.get("first", None) is None
+        ):  # pragma: no cover
```

Comment:
Same here - this is nontrivial, so I'd also like to see a test (or, if the code is never ever really used, could probably be dropped or an exception raised)

Reply:
This is now explicitly commented out since we disable it in our settings and it was impossible to test.